File: Labs/Lab02/2.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <string>

using namespace cv;
using namespace std;

// In C++, you can define constants variable using #define
#define VIDEO_FILE "../robot.qt"
#define ROTATE false

int main(int argc, char** argv)
{
    Mat matFrameCapture;
    Mat matFrameDisplay;
    int iKey = -1;

    // Open input video file
    VideoCapture videoCapture(VIDEO_FILE);
    if (!videoCapture.isOpened()) {
        cerr << "ERROR! Unable to open input video file " << VIDEO_FILE << endl;
        return -1;
    }

    namedWindow(VIDEO_FILE, WINDOW_NORMAL | WINDOW_KEEPRATIO | WINDOW_GUI_EXPANDED);
    int n_frame = 0;

    // Capture loop
    while (true)        // play video until user presses <space>
    {
        // Get the next frame
        videoCapture.read(matFrameCapture);
        if (matFrameCapture.empty())
        {
            // End of video file
            break;
        }

        // We can rotate the image easily if needed.
#if ROTATE
        rotate(matFrameCapture, matFrameDisplay, RotateFlags::ROTATE_180);   //rotate 180 degree and put the image to matFrameDisplay
#else
        matFrameDisplay = matFrameCapture;
#endif

        float ratio = 480.0 / matFrameDisplay.rows;
        resize(matFrameDisplay, matFrameDisplay, cv::Size(), ratio, ratio, INTER_LINEAR); // resize image to 480p for showing

        // Display
        imshow(VIDEO_FILE, matFrameDisplay); // Show the image in window named "robot.mp4"
        n_frame++;
        displayOverlay(VIDEO_FILE, "Frame #"+to_string(n_frame));

        iKey = -1;
        while (iKey!=int(' ') && iKey!=int('q') ) {
            iKey = waitKey(30); // Wait 30 ms to give a realistic playback speed
            if (getWindowProperty(VIDEO_FILE, WND_PROP_VISIBLE)==0) {
                   iKey = int('q');
            }
        }
        if (iKey==int('q')) {
            break;
        }

    }
    return 0;
}

File: Labs/Lab02/2.py
import cv2
import numpy as np
import sys

VIDEO_FILE = 'robot.qt'
ROTATE = False

if __name__ == '__main__':

    key = -1;

    # Open input video file
    videoCapture = cv2.VideoCapture(VIDEO_FILE);
    if not videoCapture.isOpened():
        print('Error: Unable to open input video file', VIDEO_FILE)
        sys.exit('Unable to open input video file')

    width  = videoCapture.get(cv2.CAP_PROP_FRAME_WIDTH)   # float `width`
    height = videoCapture.get(cv2.CAP_PROP_FRAME_HEIGHT)  # float `height`
    cv2.namedWindow(VIDEO_FILE, cv2.WINDOW_NORMAL | cv2.WINDOW_KEEPRATIO | cv2.WINDOW_GUI_EXPANDED)
    n_frame = 0

    # Capture loop 
    while (1):        # play video until user presses <space>
        # Get the next frame
        _, matFrameCapture = videoCapture.read()
        if matFrameCapture is None:
            # End of video
            break

        # Rotate if needed
        if ROTATE:
            _, matFrameDisplay = cv2.rotate(matFrameCapture, cv2.ROTATE_180)
        else:
            matFrameDisplay = matFrameCapture;

        ratio = 480.0 / height
        dim = (int(width * ratio), int(height * ratio))
        # resize image to 480p for display
        # matFrameDisplay = cv2.resize(matFrameDisplay, dim)

        # Show the image in window named "robot.mp4"
        cv2.imshow(VIDEO_FILE, matFrameDisplay)
        n_frame+=1
        cv2.displayOverlay(VIDEO_FILE, "Frame: "+str(n_frame))

        key = -1
        while ( key != ord(' ') and key != ord('q') ):
            key = cv2.waitKey(30)
            if cv2.getWindowProperty(VIDEO_FILE, cv2.WND_PROP_VISIBLE)==0:
                  key = ord('q')
        if key == ord('q'):
            break

